### PR TITLE
Fixed issue 18821: Public url is not used for SURVEYURL

### PR DESCRIPTION
**Pull request by twilligls**
Fixed issue #<Mantis issue number>:
New feature #<Mantis issue number>:
Dev:

### CODE REVIEW

**Comment by Shnoulle** [comment]
Another solution can be add a LSYii_Application->getBaseUrl ?

**Comment by Shnoulle** [comment]
Put this on dev mantis maybe, we need to review App->getController()->getAbsoluteUrl()

**Comment by twilligls** [comment]
> 

Doesn't work, because you would override the baseUrl all the time with configured publicurl, even when you don't need it.

**Comment by twilligls** [comment]
> Another solution can be add a LSYii_Application->getBaseUrl ?


Doesn't work, because you would override the baseUrl all the time with configured publicurl, even when you don't need it.

**Comment by twilligls** [comment]
> Put this on dev mantis maybe, we need to review App->getController()->getAbsoluteUrl()

I don't understand

**Comment by Shnoulle** [comment]
> > Put this on dev mantis maybe, we need to review App->getController()->getAbsoluteUrl()
> 
> I don't understand

Just create a Development mantis issue : https://bugs.limesurvey.org/bug_report_page.php

> > Another solution can be add a LSYii_Application->getBaseUrl ?
> 
> Doesn't work, because you would override the baseUrl all the time with configured publicurl, even when you don't need it.

 :+1: 

Maybe a new function then ?  I think we need it for API too.

**Comment by gabrieljenik** [comment]
Some comments, maybe for the DEV branch:
- I agree we need a method on the app but only for the createPublicUrl
  + Make new method based on LSYiiController::createAbsoluteUrl()
  + Make LSYiiController::createAbsoluteUrl() to call the new app method.
  + Probably rename the LSYiiController::createAbsoluteUrl() to LSYiiController::createPublicUrl()

File: application/models/Survey.php [hunk 1]
```diff
@@ -2351,7 +2357,8 @@ public function getSurveyUrl($language = null, $params = [], $preferShortUrl = t
 
         // If short url is not preferred or no alias is found, return a traditional URL
         $urlParams = array_merge($params, ['sid' => $this->sid, 'lang' => $language]);
-        $url = Yii::app()->createAbsoluteUrl('survey/index', $urlParams);
+        $url = App()->getController()->createAbsoluteUrl('survey/index', $urlParams);
```
Comment (gabrieljenik) [inline application/models/Survey.php:2360]: I don't like to call a controller from a model, but if no code reorg is to be done, I don't see any other way

Comment (gabrieljenik) [reply application/models/Survey.php:2360]: Maybe you can just move the code to a helper?
Maybe now is a good time to create App::createPublicUrl() ?

Comment (twilligls) [reply application/models/Survey.php:2360]: ok, you convinced me now :)

Comment (twilligls) [reply application/models/Survey.php]: Please check my latest commit

**Pull request merged**